### DISCUSSION TITLE
add devcontainer manager

### DISCUSSION THREAD

**Discussion by electriquo** [renovatebot/renovate#25338] [Suggest an Idea]
### Tell us more.

devcontaine adoption is growing, even [renovate project uses it](https://github.com/renovatebot/renovate/tree/main/.devcontainer). the devcontainer specification allows to [specify an image and pin its version](https://containers.dev/implementors/json_reference/#image-specific). [for instance](https://github.com/devcontainers/images/blob/b625d5ef931adb3f95a2195aafe929725c987b90/src/base-debian/README.md?plain=1#L35):

```
mcr.microsoft.com/devcontainers/base:1.0.0-bookworm
```

what do you think of adding a devcontainer manager (maybe it can leverage regex manager for that)?
could you share a location in source code where renovate reads/parses a json/yaml file and take action upon it?

**Comment by rarkins** [comment]
There are many managers which extract docker dependencies which could be used as a reference. Look for any importers of `DockerDatasource` within `lib/modules/manager/*`

**Comment by lmilbaum** [comment]
I'm using the following [configuration](https://github.com/platform-engineering-org/terraform-github-configuration-module/blob/main/renovate.json). Maybe it can help with the implementation.

**Reply by electriquo** [reply]
thanks. these configuration has hard-coded registries.
i'd love to see renovate has every this functionality out of the box.

i'll try to see how can i contribute.

**Comment by electriquo** [comment]
i took on @lmilbaum suggestion and tried to follow renovate advance capture, then i noticed that:
1. [advance capture](https://docs.renovatebot.com/modules/manager/regex/#advanced-capture) does not capture the registry when it is given. for instance, it will work nicely for
   ```dockerfile
   IMAGE=node:12@sha256:6e5264cd4cfaefd7174b2bc10c7f9a1c2b99d98d127fc57a802d264da9fb43bd
   ```
   but misses (the optional [registryUrls](https://docs.renovatebot.com/configuration-options/#registryurls), which probably should have docker.io as it default value) when a fully qualified image name is used
   ```dockerfile
   IMAGE=docker.io/library/node:12@sha256:6e5264cd4cfaefd7174b2bc10c7f9a1c2b99d98d127fc57a802d264da9fb43bd
   ```
   here is a [miss example](https://regex101.com/r/aKFnLk/1) and a [fix example](https://regex101.com/r/ZoQO3A/1).
   it seems that there is no need to [update renovate to update Dockerfile as the document explains](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/custom/regex/readme.md?plain=1#L100), as it seems that it is supported out of the box when using [builtin presets, i just couldn't find which one](https://github.com/electriquo/ren-docker/issues/1).
   @rarkins could you point where in the codebase this take place so maybe i could fix it?
2. i wanted to apply the same logic above to the image within the devcontainer.json file, but strangely, i had to modify slightly [the regular expression used above](https://regex101.com/r/ZoQO3A/1) to achieve a [new regular expression](https://regex101.com/r/2xlt44/1), but i couldn't figure out why it had to be modified (excluding the regex prefix and suffix). can someone please explain?

**Reply by rarkins** [reply]
I don't understand what you're asking me to fix

**Reply by electriquo** [reply]
not you, i'll do it but i need a pointer to the codebase where renovate uses the regular expression to manage updates for the docker image in Dockerfile
>  @rarkins could you point where in the codebase this take place so maybe i could fix it?

plus, any idea why the regex that extract the same string had to be modified?

**Reply by HonkingGoose** [reply]
@TWiStErRob do you want to help with a regex question? See the comments above.

**Reply by electriquo** [reply]
@rarkins i think we should convert (1) above to an issue/bug. what do you think?

**Reply by rarkins** [reply]
I think we already have one: #12116

**Reply by TWiStErRob** [reply]
@electriquo re:
> why the regex that extract the same string had to be modified?

As far as I see you made the `@<digest>` part optional. I think the original "Advanced capture" example is simply demonstrating how to write regex-es in Renovate, and capture the groups e.g. `depName`, and therefore it's not a complete pattern to match all types of Docker image definitions (e.g. registry, optional digest, different digest algorithm, etc.) It's probably kept simple for the documentation.

I think in this case you're developing a generic reusable piece of code/regex (not meant for alteration), so you need to figure out what the possible formats can be. https://docs.docker.com/engine/reference/commandline/pull/ and https://docs.docker.com/engine/reference/builder/#from are probably good starting point, but they don't look like scientifically well-defined patterns. Although there must be some "standard" grammar for image names/references/uris/whatever they are.

---

Sidenote: instead of making two optional things, it's probably better to have just one:
`@?(?<currentDigest>sha256:[a-f0-9]+)?` &rarr; `(@(?<currentDigest>sha256:[a-f0-9]+))?` because otherwise `foo/bar@` is a "valid" pattern. From my limited knowledge about containers (and based on [docs](https://docs.docker.com/engine/reference/builder/#from)), I think the `@` and the digest "belong together", similarly how `:` and tag do.

**Comment by electriquo** [comment]
> I think we already have one: https://github.com/renovatebot/renovate/issues/12116

@rarkins  was referring not to the devcontainer but to the version update of the docker image within the dockerfile that does not take into account the docker registry.

**Reply by rarkins** [reply]
See the comment from @TWiStErRob above. It's not intended to solve every use case, and is not a bug

**Reply by electriquo** [reply]
> It's not intended to solve every use case

would you please share why is it preferable for renovate to handle a specific use case and not to handle the generic case with something as [as this](https://regex101.com/r/iQAmGM/1)?

**Reply by rarkins** [reply]
No, but if you want to propose an improvement which is backwards-compatible, go ahead
